File: packages/lynx/host/android/app/src/main/java/com/yee94/openchamber/lynx/OpenChamberLynxCameraAdapter.kt
package com.yee94.openchamber.lynx

/** Cap BarcodeScanner spirit — without CameraX this is unavailable (never fake pairing). */
sealed class LynxCameraScanResult {
    data class Ok(val rawValue: String) : LynxCameraScanResult()
    object Cancelled : LynxCameraScanResult()
    object PermissionDenied : LynxCameraScanResult()
    object Unavailable : LynxCameraScanResult()
    data class Failed(val error: String) : LynxCameraScanResult()
}

interface LynxCameraScanning {
    suspend fun scanPairingQr(): LynxCameraScanResult
}

/**
 * Stub binder. Host replaces with CameraX / ML Kit barcode when shipping a device binary.
 * Callback path: scanPairingQr → Bridge.QrScanResult → Lynx parseConnectionPayload.
 */
class OpenChamberLynxCameraAdapter(
    private val implementation: (suspend () -> LynxCameraScanResult)? = null,
) : LynxCameraScanning {
    override suspend fun scanPairingQr(): LynxCameraScanResult {
        if (implementation != null) return implementation.invoke()
        // Inject point: CameraX analyzer → return Ok(rawValue = payload)
        return LynxCameraScanResult.Unavailable
    }
}

File: packages/lynx/host/ios/OpenChamberLynxCameraAdapter.swift
import Foundation

/// Cap BarcodeScanner spirit for Lynx 扫一扫.
/// Without AVCaptureSession wiring this returns `.unavailable` — never fake a pairing payload.
/// Successful scans emit through OpenChamberLynxBridge → Lynx pairing parse.
enum OpenChamberLynxCameraScanResult {
  case ok(rawValue: String)
  case cancelled
  case permissionDenied
  case unavailable
  case failed(String)
}

protocol OpenChamberLynxCameraScanning {
  func scanPairingQr() async -> OpenChamberLynxCameraScanResult
}

/// Stub binder. Host replaces with AVCaptureMetadataOutput / Vision when shipping a device binary.
/// Callback path: scanPairingQr → Bridge.qrScanResult → Lynx `parseConnectionPayload`.
final class OpenChamberLynxCameraAdapter: OpenChamberLynxCameraScanning {
  /// Optional inject for tests / host camera module.
  var implementation: (() async -> OpenChamberLynxCameraScanResult)?

  func scanPairingQr() async -> OpenChamberLynxCameraScanResult {
    if let implementation {
      return await implementation()
    }
    // Inject point: present AVCaptureSession QR scanner; return .ok(rawValue: payload).
    return .unavailable
  }
}
